src/c++/perf_analyzer/client_backend/openai/http_client.cc

```diff
@@ -183,94 +183,102 @@ HttpClient::SetSSLCurlOptions(CURL* curl_handle)
-          // This shouldn't happen
-          if (itr == ongoing_async_requests_.end()) {
-            std::cerr
-                << "Unexpected error: received completed request that is not "
-                   "in the list of asynchronous requests"
+    mc = curl_multi_poll(multi_handle_, NULL, 0, 1000, &numfds);
+    if (mc != CURLM_OK) {
+      std::cerr << "Unexpected error: curl_multi_poll failed. Code:" << mc
                 << std::endl;
-            curl_multi_remove_handle(multi_handle_, msg->easy_handle);
-            curl_easy_cleanup(msg->easy_handle);
-            continue;
-          }
-
-          uint32_t http_code = 400;
-          if (msg->data.result == CURLE_OK) {
-            curl_easy_getinfo(
-                msg->easy_handle, CURLINFO_RESPONSE_CODE, &http_code);
-          } else if (msg->data.result == CURLE_OPERATION_TIMEDOUT) {
-            http_code = 499;
-          }
-
-          request_list.emplace_back(std::move(itr->second));
-          ongoing_async_requests_.erase(itr);
-          curl_multi_remove_handle(multi_handle_, msg->easy_handle);
-          curl_easy_cleanup(msg->easy_handle);
-
-          std::unique_ptr<HttpRequest>& async_request = request_list.back();
-          async_request->http_code_ = http_code;
-
-          if (msg->msg != CURLMSG_DONE) {
-            // Something wrong happened.
-            std::cerr << "Unexpected error: received CURLMsg=" << msg->msg
-                      << std::endl;
-          }
-        }
-      } else {
-        std::cerr << "Unexpected error: curl_multi failed. Code:" << mc
+      continue;
+    }
+
+    lock.lock();  // Re-acquire the mutex before processing responses
+    while ((msg = curl_multi_info_read(multi_handle_, &place_holder))) {
+      uintptr_t identifier = reinterpret_cast<uintptr_t>(msg->easy_handle);
+      auto itr = ongoing_async_requests_.find(identifier);
+      // This shouldn't happen
+      if (itr == ongoing_async_requests_.end()) {
+        std::cerr << "Unexpected error: received completed request that is not "
+                  << "in the list of asynchronous requests" << std::endl;
+        curl_multi_remove_handle(multi_handle_, msg->easy_handle);
+        curl_easy_cleanup(msg->easy_handle);
+        continue;
+      }
+
+      uint32_t http_code = 400;
+      if (msg->data.result == CURLE_OK) {
+        curl_easy_getinfo(msg->easy_handle, CURLINFO_RESPONSE_CODE, &http_code);
+      } else if (msg->data.result == CURLE_OPERATION_TIMEDOUT) {
+        http_code = 499;
+      }
+
+      request_list.emplace_back(std::move(itr->second));
+      ongoing_async_requests_.erase(itr);
+      curl_multi_remove_handle(multi_handle_, msg->easy_handle);
+      curl_easy_cleanup(msg->easy_handle);
+
+      std::unique_ptr<HttpRequest>& async_request = request_list.back();
+      async_request->http_code_ = http_code;
+
+      if (msg->msg != CURLMSG_DONE) {
+        std::cerr << "Unexpected error: received CURLMsg=" << msg->msg
```

Comment:
Unsure where this is coming from yet but in your testing, did you see this message pop up?

Reply:
Nope, didn't come up once.